Broken pattern span for identifier parameters in macros

This kind of macro pattern seems to mess up the span:

``` rust
macro_rules! bad {
    ($s:ident whatever) => {
        {
            let $s = 0;
            *&mut $s = 0;
        }
    }
}

fn main() {
    bad!(foo whatever);
}
```

[Playpen](http://is.gd/ZtC5C0)

Compiling it results in a messed up error message:

```
<anon>:11:14: 11:22 error: cannot borrow immutable local variable `foo` as mutable
<anon>:11     bad!(foo whatever);
                       ^~~~~~~~
<anon>:11:5: 11:24 note: in this expansion of bad! (defined in <anon>)
<anon>:11:14: 11:22 help: to make the local variable mutable, use `mut` as shown:
<anon>:       bad!(foo mut whatever);
<anon>:11:5: 11:24 note: in this expansion of bad! (defined in <anon>)
error: aborting due to previous error
```

This is almost certainly not what we want today.. Or at least, it looks really odd.

```
error[E0596]: cannot borrow immutable local variable `foo` as mutable
  --> test.rs:11:10
   |
5  | |             *&mut $s = 0;
   | |____________________^ cannot borrow mutably
...
11 |       bad!(foo whatever);
   |  _____-----^-------------
   | |     |
   | |     in this macro invocation

error: aborting due to previous error(s)
```

The current output is beyond bad because of the new suggestion:

```
error[E0596]: cannot borrow immutable local variable `foo` as mutable
  --> src/main.rs:5:19
   |
4  |                let $s = 0;
   |   _________________-
5  |  |             *&mut $s = 0;
   |  |___________________^
6  | ||         }
7  | ||     }
8  | || }
9  | ||
10 | || fn main() {
11 | ||     bad!(foo whatever);
   | ||     -------^-----------
   | ||_____|______|
   | |______|______consider changing this to `mut $s = 0;
            *&mut $s = 0;
        }
    }
}

fn main() {
    bad!(foo`
   |        |      cannot borrow mutably
   |        in this macro invocation
```

Fixed in #47942.

--- a/src/libsyntax_pos/lib.rs
+++ b/src/libsyntax_pos/lib.rs
@@ -361,13 +361,24 @@ impl Span {
 
     /// Return a `Span` that would enclose both `self` and `end`.
     pub fn to(self, end: Span) -> Span {
-        let span = self.data();
-        let end = end.data();
+        let span_data = self.data();
+        let end_data = end.data();
+        // FIXME(jseyfried): self.ctxt should always equal end.ctxt here (c.f. issue #23480)
+        // Return the macro span on its own to avoid weird diagnostic output. It is preferable to
+        // have an incomplete span than a completely nonsensical one.
+        if span_data.ctxt != end_data.ctxt {
+            if span_data.ctxt == SyntaxContext::empty() {
+                return end;
+            } else if end_data.ctxt == SyntaxContext::empty() {
+                return self;
+            }
+            // both span fall within a macro
+            // FIXME(estebank) check if it is the *same* macro
+        }
         Span::new(
-            cmp::min(span.lo, end.lo),
-            cmp::max(span.hi, end.hi),
-            // FIXME(jseyfried): self.ctxt should always equal end.ctxt here (c.f. issue #23480)
-            if span.ctxt == SyntaxContext::empty() { end.ctxt } else { span.ctxt },
+            cmp::min(span_data.lo, end_data.lo),
+            cmp::max(span_data.hi, end_data.hi),
+            if span_data.ctxt == SyntaxContext::empty() { end_data.ctxt } else { span_data.ctxt },
         )
     }
 

--- a/src/test/ui/macros/span-covering-argument-1.rs
+++ b/src/test/ui/macros/span-covering-argument-1.rs
@@ -1,0 +1,23 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+macro_rules! bad {
+    ($s:ident whatever) => {
+        {
+            let $s = 0;
+            *&mut $s = 0;
+            //~^ ERROR cannot borrow immutable local variable `foo` as mutable [E0596]
+        }
+    }
+}
+
+fn main() {
+    bad!(foo whatever);
+}

--- a/src/test/ui/macros/span-covering-argument-1.stderr
+++ b/src/test/ui/macros/span-covering-argument-1.stderr
@@ -1,0 +1,13 @@
+error[E0596]: cannot borrow immutable local variable `foo` as mutable
+  --> $DIR/span-covering-argument-1.rs:15:19
+   |
+14 |             let $s = 0;
+   |                 -- consider changing this to `mut $s`
+15 |             *&mut $s = 0;
+   |                   ^^ cannot borrow mutably
+...
+22 |     bad!(foo whatever);
+   |     ------------------- in this macro invocation
+
+error: aborting due to previous error
+

--- a/src/test/ui/span/macro-span-replacement.stderr
+++ b/src/test/ui/span/macro-span-replacement.stderr
@@ -1,8 +1,8 @@
 warning: struct is never used: `S`
-  --> $DIR/macro-span-replacement.rs:17:9
+  --> $DIR/macro-span-replacement.rs:17:14
    |
 17 |         $b $a; //~ WARN struct is never used
-   |         ^^^^^^
+   |              ^
 ...
 22 |     m!(S struct);
    |     ------------- in this macro invocation

